bin/coqtop.ml

```diff
@@ -0,0 +1,123 @@
+
+let term =
+  let+ common = Common.term
+  and+ context =
+    Common.context_arg ~doc:{|Run the Coq toplevel in this build context.|}
+  and+ coqtop =
+    let doc = "Run the given toplevel command instead of the default." in
+    Arg.(value & opt string "coqtop" & info [ "toplevel" ] ~docv:"CMD" ~doc)
+  and+ coq_file_arg =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        (* Checks on the argument file. *)
+        let source = Common.prefix_target common coq_file_arg in
+        if not (Path.is_file (Path.of_string source)) then
+          User_error.raise [ Pp.textf "cannot find file: %s" coq_file_arg ];
+        let coq_file_arg =
+          let cwd = Path.external_ (Path.External.initial_cwd) in
+          let file = Path.of_filename_relative_to_initial_cwd coq_file_arg in
+          let cwd = Path.to_string cwd in
+          let file = Path.to_string file in
+          match String.drop_prefix ~prefix:(cwd ^ "/") file with
+          | None -> coq_file_arg
+          | Some s -> s
+        in
+        let coq_file_build =
+          let p = Common.prefix_target common coq_file_arg in
+          Path.Build.relative context.build_dir p
+        in
+        let dir =
+          let dir = Filename.dirname coq_file_arg in
+          let p = Common.prefix_target common dir in
+          Path.Build.relative context.build_dir p
+        in
+        let* (coqtop, args) =
+          let open Memo.Build.O in
+          Build_system.run_exn (fun () ->
+              let* dc = Dune_rules.Dir_contents.get sctx ~dir in
+              let* coq_src = Dune_rules.Dir_contents.coq dc in
+              let coq_module =
+                let source = coq_file_build in
+                match Dune_rules.Coq_sources.find_module ~source coq_src with
+                | Some m -> snd m
+                | None -> User_error.raise
+                            [ Pp.textf "cannot find file: %s" coq_file_arg ]
+              in
+              let* _ : unit * Dep.Fact.t Dep.Map.t =
+                let deps =
+                  Dune_rules.Coq_rules.deps_of ~dir ~boot_type:No_boot coq_module
+                in
+                Action_builder.run deps Eager
+              in
+              let stanza =
+                let name = Dune_rules.Coq_module.name coq_module in
+                Dune_rules.Coq_sources.lookup_module coq_src name
+              in
+              let* args =
+                match stanza with
+                | None -> Memo.Build.return [] (* FIXME? *)
```

Comment:
What about just erroring in this case? Something like "The <file> does not belong to any theory".

Reply:
Thanks for the comments! I'm actually not sure if giving an error is the best option here. My thinking was that a user might want to step through a stand-alone file that is not yet part of any theory. If the user tries to import anything from another theory then they'll get an error from `coqtop`, and they'll know that they need to make the dependencies available. @ejgallego any opinion on that?

Reply:
I imagine the following situation: a user adds a `.v` file to a directory and forgets to include it in an appropriate theory. Then the user starts developing with coqtop and is confused by the absence of libraries for example. It might take a while to figure out what is actually happening. Not sure how realistic this scenario is, but on the OCaml side, we forbid modules from being used unless they're part of a stanza.

Reply:
Yeah, that makes sense. My worry was that users might be annoyed if they open such a new file in their editor, and `coqtop` just fails right away. On the other hand, in most cases, opening a new file will just work provided it is placed in a directory with an appropriate theory. One question remains though: what if the file has been opened, but never saved? In that case, dune will not know of that file at all, right? In fact, there is the same problem if you add a new dependency to a file: the only way for the system to take it into account is to save and restart the toplevel. Not sure if that behaviour is what we want in practice.

Reply:
> One question remains though: what if the file has been opened, but never saved? In that case, dune will not know of that file at all, right? In fact, there is the same problem if you add a new dependency to a file: the only way for the system to take it into account is to save and restart the toplevel. Not sure if that behaviour is what we want in practice.

IMO, it's fine to ask users to save the file. It's not pretty, but it's consistent with how the rest of the dune integration works. Basically, it's one of those heuristics that's easy to remember for users and can be applied in more than one place. As opposed to stuff like "if the module doesn't belong to a theory, we will assume no dependencies or flags".

Reply:
OK, I've applied your suggestion and extended the documentation to highlight the limitations.